### PR TITLE
Infrastructure for using channel-dependent pulse shapes for HB/HE

### PR DESCRIPTION
**Pull request by igv4321**
#### PR description:

Adding infrastructure for modeling Hcal front-end pulse shapes channel-by-channel. The pulse shapes were determined by phase scans, as described in https://indico.cern.ch/event/1336438/contributions/5626382/attachments/2733633/4753627/HBHE_Pulse_shape_study.pdf. The new pulse shapes have not yet been turned on, pending final timing adjustment.

#### PR validation:

The usual matrix text were run. Also, privately, the new pulse shapes were switched on (parameter "useChannelShapes" in HBHEPhase1Reconstructor_cfi.py was set True), and the code appears to produce reasonable results on the 2023 data workflow, as indicated by the subsequent Hcal rechit dump.

### CODE REVIEW

**Comment by jfernan2** [comment]
please test

File: RecoLocalCalo/HcalRecProducers/python/HBHEPhase1Reconstructor_cfi.py [hunk 1]
```diff
@@ -9,6 +9,11 @@
 hbheprereco = cms.EDProducer(
     "HBHEPhase1Reconstructor",
 
+    # Label for the HcalPulseShapeLookup object and flag
+    # indicating whether we are using it
+    channelShapesLabel = cms.string("HcalDataShapes"),
+    useChannelShapes = cms.bool(False),
+
```
Comment (missirol) [inline RecoLocalCalo/HcalRecProducers/python/HBHEPhase1Reconstructor_cfi.py:16]: Hi, I'm not a reviewer, I just have a general question on these developments.

What is their target (if there is one) for going in production ? 2025 data-taking ?

Are these changes only needed for the HCAL offline reconstruction, or are they also supposed to be used at HLT ?

Ever since the heterogeneous version of the HCAL local reconstruction was ported to Alpaka (https://github.com/cms-sw/cmssw/pull/44910), the plugin `HBHEPhase1Reconstructor` is not used anymore in the Run-3 HLT menus (in other words, HLT is blind to this PR, as far as I understand).

Comment (igv4321) [reply RecoLocalCalo/HcalRecProducers/python/HBHEPhase1Reconstructor_cfi.py:16]: This will be useful for 2023 and 2024 re-reco as well as for the future data taking. Unfortunately, at the moment we don't have an alpaka expert in the HCAL DPG, so there is no concrete plan how to get this to work in the HLT. Of course, help from Martin Kwok (or any other alpaka expert) would be appreciated.

Comment (missirol) [reply RecoLocalCalo/HcalRecProducers/python/HBHEPhase1Reconstructor_cfi.py:16]: Thanks for the clarification.

Notifying @kakwok, @fwyzard and @cms-sw/hlt-l2.

Comment (missirol) [reply RecoLocalCalo/HcalRecProducers/python/HBHEPhase1Reconstructor_cfi.py:16]: I have to say that this

> for the future data taking

combined with

> there is no concrete plan how to get this to work in the HLT.

sounds to me like the HCAL DPG is not taking full ownership of the HCAL code running in production today in the HLT. I don't know the quantitative impact of these channel-dependent pulse shapes, but I would assume that we (well, HCAL specifically) should make an effort to keep HLT and offline as close as possible to each other.

File: RecoLocalCalo/HcalRecAlgos/interface/MahiFit.h [hunk 2]
```diff
@@ -96,7 +96,7 @@ struct MahiDebugInfo {
 class MahiFit {
 public:
   MahiFit();
-  ~MahiFit() {}
+  ~MahiFit() {};
```
Comment (fwyzard) [inline RecoLocalCalo/HcalRecAlgos/interface/MahiFit.h:99]:
```suggestion
  ~MahiFit() {}
```

File: RecoLocalCalo/HcalRecAlgos/python/hcalPulseShapesESProd_cfi.py [hunk 3]
```diff
@@ -0,0 +1,12847 @@
...
+    mapDumpFile = cms.untracked.string(''),
+    pulseShapeLength = cms.uint32(250),
+    globalTimeShift = cms.int32(-75),
+    defaultTimeShift = cms.int32(0),
+    pulseShapes = cms.VPSet(
+        cms.PSet(
+            label = cms.string('default'),
+            t0 = cms.uint32(0),
+            pulse = cms.vdouble(
+                5.22174e-12, 7.04852e-10, 3.49584e-08, 7.78029e-07, 9.11847e-06,
```
Comment (mmusich) [inline RecoLocalCalo/HcalRecAlgos/python/hcalPulseShapesESProd_cfi.py:15]: wouldn't this large amount of parameters be better stored in the database, or at least in some external `cms-data` repository? This makes this file ~ unreadable.

Comment (igv4321) [reply RecoLocalCalo/HcalRecAlgos/python/hcalPulseShapesESProd_cfi.py:15]: Yes, this was discussed with the Hcal calibration group. However, this is the first time we are introducing this channel-by-channel dependence since the beginning of CMS operations, and the pulse shapes are not expected to change more often than once per era, so it was decided to keep the information in a config file. While this file is large (and, of course, computer-generated), its structure is rather simple, as can be seen from the "fillDescriptions" function of the HcalPulseShapesEP.cc code.

Comment (fwyzard) [reply RecoLocalCalo/HcalRecAlgos/python/hcalPulseShapesESProd_cfi.py:15]: This approach is not acceptable.

Please move the pulse shapes either to a text file in an external `cms-data` package, or to a database payload.

Comment (fwyzard) [reply RecoLocalCalo/HcalRecAlgos/python/hcalPulseShapesESProd_cfi.py:15]: assign core

Comment (fwyzard) [reply RecoLocalCalo/HcalRecAlgos/python/hcalPulseShapesESProd_cfi.py:15]: assign alca

**Comment by fwyzard** [comment]
assign alca

Please review the misuse of python configurations for detector calibrations, _e.g._ in `RecoLocalCalo/HcalRecAlgos/python/hcalPulseShapesESProd_cfi.py`.

File: RecoLocalCalo/HcalRecAlgos/python/hcalPulseShapesESProd_cfi.py (hunk 3 above)
Comment (igv4321) [reply RecoLocalCalo/HcalRecAlgos/python/hcalPulseShapesESProd_cfi.py:15]: Could you please substantiate your statement? Haven't seen anything against this approach in CMSSW guidelines and can't understand why it can be harmful.

**Comment by igv4321** [comment]
Perhaps, I should point out that these are not "detector calibrations" in the usual sense of the word. Pulse shapes can change, but the expected frequency is about the same as the frequency with which the detector composition is modified (once per shutdown), or front end electronics and/or firmware are swapped. The "eras" were invented precisely to permit such infrequent modifications in a reasonable way.

File: RecoLocalCalo/HcalRecAlgos/python/hcalPulseShapesESProd_cfi.py (hunk 3 above)
Comment (fwyzard) [reply RecoLocalCalo/HcalRecAlgos/python/hcalPulseShapesESProd_cfi.py:15]: Have you seen many other modules with a configuration that takes **twelve thousand lines** of python code ?

**Comment by fwyzard** [comment]
> Perhaps, I should point out that these are not "detector calibrations" in the usual sense of the word. Pulse shapes can change, but the expected frequency is about the same as the frequency with which the detector composition is modified (once per shutdown), or front end electronics and/or firmware are swapped. The "eras" were invented precisely to permit such infrequent modifications in a reasonable way.

Perfect, so they can be put in a database object that will not need to be updated frequently.

**Comment by igv4321** [comment]
Yes, a database object can be developed, but that requires an extra level of complexity. What are the advantages of it?

**Comment by missirol** [comment]
> What are the advantages of it?

Some examples.

 - Not tying these conditions to a specific CMSSW release and/or HLT menu.

 - (Assuming this would be eventually used at HLT) Not adding O(13k) lines of python code to every HLT menu that needs to be deployed at P5 for a single set of HCAL calibration constants (not to mention that, if this was added in this form in the HLT menu, it would then clog ConfDB, i.e. the db where HLT menus are stored, and it would get copied over hundreds of development menus created by users across CMS in the months ahead).

 - Making it (arguably) easier to maintain and inspect them via standard conditions-db tools.

 - Making it easier to know when they were deployed online.

 - Making it possible to load them in the appropriate IOVs when re-running on old data with newer CMSSW releases and/or newer HLT menus.

**Comment by fwyzard** [comment]
That assumes that the python-based approach would stay, and a database approach would be added.
But the python-based approach should not be there at all.

So, a database approach does not require an "extra_ level of complexity", it only requires a different approach.

As to why adding a 1 MB python file with 50'000 parameters should be discouraged:
  - overall size: just like large data files that rarely change should not be put in the main repository (see https://cms-sw.github.io/cms_coding_rules.html#data-files);
  - efficiency: every single `cmsRun` job will have to parse the same parameters over and over again;
  - memory usage: have you measured how much memory does the resulting `ParameterSet` take ?
  - unsuitable for online: due to the way they are stored and versioned, HLT configurations are fully expanded; adding so many more parameters to it would be extremely wasteful, and something HLT L2 have already commented against.

To put it simply: this is not a configuration, this is data, and it should be treated accordingly.

**Comment by abdoulline** [comment]
As a temporary "get though" workaround [which was mentioned by Andrea above](https://github.com/cms-sw/cmssw/pull/45995#discussion_r1759744303) and which I've suggested ca a year ago - 
to ask Shahzad to put this "file of interest" into the external cms-data ?  🤔

**Comment by mmusich** [comment]
> As a temporary "get though" workaround 

Indeed this would avoid some of nastiest consequences of having a 12k lines python configuration fragment to load at runtime as well as making this slightly more maintainable (Marino and Andrea have given reasons enough above why this is really undesirable). 
Allow me to remind that:
   * in order for this to be efficient you would need anyway to ship the parameters for execution via an EventSetup object. This requires definition of an ES class / record pair, independently of persisting in the DB;
   * with this "external" approach every "calibration" update is tied to a cmssw version change - with all the operational overhead it implies;
   * other subsystems have developed DB objects for similar purposes in the past;
   * "get-through" solutions generally tend to stick around indefinitely.

**Comment by perrotta** [comment]
@igv4321 since you are developing this part of the code anew please consider using the condition database option that was suggested, whose advantage has been thoroughly detailed here above.
For a support, you can find the names of the AlCaDB contacts in HCAL from this list: https://twiki.cern.ch/twiki/bin/viewauth/CMS/AlCaDBContacts

**Comment by abdoulline** [comment]
>  For a support, you can find the names of the AlCaDB contacts in HCAL from this list: https://twiki.cern.ch/twiki/bin/viewauth/CMS/AlCaDBContacts

I'm afraid HCAL (three) names are obsolete. Yildiray Komurgu to replace them.

**Comment by fwyzard** [comment]
type hcal

**Comment by igv4321** [comment]
Superseeded by #49556
